### PR TITLE
plan/typeinferer: fix a bug when mathematic expression contains datetime or duration operands

### PR DESCRIPTION
**Pull request by XuHuaiyu**
fix issue #2593 
When a mathematic expression contains operands with datetime or duration type,
the type of the expression would be inferred as mysql.TypeLongLong by mistake.
e.g. `now() + 2` , `curtime()+2`.

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@coocood @zimulala @hanfei1991 @tiancaiamao PTAL

**Comment by coocood** [comment]
We need to make sure MySQL returns the same type.
One way is to use MySQLWorkbench, there is a `FieldTypes` tab, you can get the type of the result field.

**Comment by XuHuaiyu** [comment]
@coocood PTAL

**Comment by XuHuaiyu** [comment]
That's ok, but I think it's may be better to change it in an individual PR。 @ngaut

**Comment by ngaut** [comment]
Sure

**Comment by XuHuaiyu** [comment]
@ngaut I have a thought again, it may be better to keep Tp public.
Tp is a field of struct `FieldType`, and there are another 6 public fields in FieldType.
If we make `Tp` private, we would better to make the other fields private, and add set/get methods for every field for using them in other packages. I think it's ***not really needed*** to make the fields private and add the set/get methods.

**Comment by XuHuaiyu** [comment]
@coocood @zimulala @tiancaiamao PTAL

File: plan/typeinferer.go [hunk 1]
```diff
@@ -192,7 +192,34 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
 	x.Type.Collate = charset.CollationBin
 }
 
-func mergeArithType(a, b byte) byte {
+func mergeArithType(tpa, tpb *types.FieldType) byte {
+	a, b := tpa.Tp, tpb.Tp
+	isDatetimeOrDuration := false
+	if a == mysql.TypeDatetime || a == mysql.TypeDuration {
```
Comment (coocood) [inline plan/typeinferer.go]: Should consider `mysql.TypeTimestamp`?

File: plan/typeinferer.go [hunk 2]
```diff
@@ -192,7 +192,34 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
...
+		isDatetimeOrDuration = true
+		if tpa.Decimal != 0 {
+			a = mysql.TypeDouble
```
Comment (coocood) [inline plan/typeinferer.go]: Should be `mysql.NewDecimal`?

File: plan/typeinferer.go [hunk 3]
```diff
@@ -192,7 +192,34 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
...
+		isDatetimeOrDuration = true
+		if tpb.Decimal != 0 {
+			b = mysql.TypeDouble
+		} else {
+			b = mysql.TypeLonglong
+		}
+	}
+	tp := mergeType(a, b)
+	if isDatetimeOrDuration && tp != mysql.TypeLonglong {
+		tp = mysql.TypeDecimal
```
Comment (coocood) [inline plan/typeinferer.go]: `mysql.TypeNewDecimal`

File: plan/typeinferer.go [hunk 4]
```diff
@@ -220,7 +247,7 @@ func (v *typeInferrer) unaryOperation(x *ast.UnaryOperationExpr) {
 		x.Type.Init(mysql.TypeLonglong)
 		if x.V.GetType() != nil {
 			switch x.V.GetType().Tp {
-			case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat:
+			case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat, mysql.TypeDatetime, mysql.TypeDuration:
```
Comment (coocood) [inline plan/typeinferer.go]: Missing `mysql.TypeTimestamp`?

File: plan/typeinferer_test.go [hunk 5]
```diff
@@ -73,6 +75,18 @@ func (ts *testTypeInferrerSuite) TestInferType(c *C) {
 
 		{"1 + '1'", mysql.TypeDouble, charset.CharsetBin},
 		{"1 + 1.1", mysql.TypeNewDecimal, charset.CharsetBin},
+		{"now() + 0", mysql.TypeLonglong, charset.CharsetBin},
+		{"curtime() + 0", mysql.TypeLonglong, charset.CharsetBin},
+		{"now(0) + 0", mysql.TypeLonglong, charset.CharsetBin},
+		{"now(2) + 0", mysql.TypeNewDecimal, charset.CharsetBin},
+		{"now() + 1.1", mysql.TypeNewDecimal, charset.CharsetBin},
+		{"now() + '1'", mysql.TypeNewDecimal, charset.CharsetBin},
```
Comment (coocood) [inline plan/typeinferer_test.go]: `now() + '1'` should be double.

File: plan/typeinferer.go [hunk 6]
```diff
@@ -192,7 +192,44 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
...
+			isDatetimeOrDuration = true
+			if tpb.Decimal > 0 {
+				b = mysql.TypeNewDecimal
+			} else {
+				b = mysql.TypeLonglong
+			}
+		}
+	}
+	tp := mergeType(a, b)
+	if isDatetimeOrDuration && tp != mysql.TypeLonglong {
```
Comment (coocood) [inline plan/typeinferer.go]: This can be removed?

File: plan/typeinferer.go [hunk 7]
```diff
@@ -192,7 +192,44 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
...
+			if tpa.Decimal > 0 {
+				a = mysql.TypeNewDecimal
+			} else {
+				a = mysql.TypeLonglong
+			}
+		}
+	}
+	if b == mysql.TypeDatetime || b == mysql.TypeDuration || b == mysql.TypeTimestamp {
+		switch a {
+		case mysql.TypeString, mysql.TypeVarString, mysql.TypeVarchar:
```
Comment (coocood) [inline plan/typeinferer.go]: Can be removed?

**Comment by coocood** [comment]
How about defining a function that converts `FieldType` to single byte type for merge arith type?
So `mysql.TypeDatetime` will be convert to `mysql.TypeNewDecimal` if `Decimal` is not zero.

File: plan/typeinferer.go [hunk 8]
```diff
@@ -192,7 +192,44 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
 	x.Type.Collate = charset.CollationBin
 }
 
-func mergeArithType(a, b byte) byte {
+func mergeArithType(tpa, tpb *types.FieldType) byte {
+	a, b := tpa.Tp, tpb.Tp
+	isDatetimeOrDuration := false
+	if a == mysql.TypeDatetime || a == mysql.TypeDuration || a == mysql.TypeTimestamp {
```
Comment (zimulala) [inline plan/typeinferer.go]: It can be extracted into a function like `IsTypeBlob` in `util/types/etc.go`

File: plan/typeinferer.go [hunk 9]
```diff
@@ -192,7 +192,19 @@ func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
...
 }
 
-func mergeArithType(a, b byte) byte {
+// toArithType converts DateTime, Duration and Timestamp types to NewDecimal type if Decimal > 0.
+func toArithType(ft *types.FieldType) (tp byte) {
+	tp = ft.Tp
+	if types.IsTypeTime(tp) {
+		if ft.Decimal > 0 {
+			tp = mysql.TypeNewDecimal
+		}
```
Comment (coocood) [inline plan/typeinferer.go]: If decimal == 0, should return mysql.TypeLonglong

File: plan/typeinferer.go [hunk 10]
```diff
@@ -279,7 +291,7 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 		if len(x.Args) > 0 {
 			tp = x.Args[0].GetType()
 			for i := 1; i < len(x.Args); i++ {
-				mergeArithType(tp.Tp, x.Args[i].GetType().Tp)
+				mergeArithType(tp, x.Args[i].GetType())
```
Comment (coocood) [inline plan/typeinferer.go]: This should be fixed in another PR.

Comment (XuHuaiyu) [inline plan/typeinferer.go]: ok

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM
